nibabel/loadsave.py

```diff
@@ -36,6 +36,11 @@ def load(filename, **kwargs):
     img : ``SpatialImage``
        Image of guessed type
     '''
+    if hasattr(filename, '__fspath__'):
+        filename = filename.__fspath__()
+    else:
+        filename = unicode(filename)
```

Comment:
I'm not sure that this is the best way to handle this. My reading is that Windows uses `unicode` paths, while Unix uses `bytes` paths. While this may work most of the time, I don't see any value to modifying the type of `filename` where we haven't, before.

As suggested by @fepegar, this may be a better way to approach this:

```Python
if isinstance(filename, getattr(os, "PathLike", ())):
    filename = os.fspath(filename)
```

As this effectively tests for whether `pathlib` exists and then allows it to handle any type conversions needed. I would be inclined to factor this out into a utility function:

```Python
def get_fname(pathlike):
    if isinstance(pathlike, getattr(os, "PathLike", ())):
        pathlike = os.fspath(pathlike)
    return pathlike
```

This could then be called in `load`, `save`, [`FileBasedImage.from_filename()`](https://github.com/nipy/nibabel/blob/19fd3f82aead1f4fe1f45ac13139acc3ad30140f/nibabel/filebasedimages.py#L257-L260) and [`FileBasedImage.to_filename()`](https://github.com/nipy/nibabel/blob/19fd3f82aead1f4fe1f45ac13139acc3ad30140f/nibabel/filebasedimages.py#L319-L334), to fully support `pathlib` objects.

Reply:
Thinking a little bit more, a simpler alternative would be something like the following:

```Python
try:
    from os import fspath
except ImportError:
    def fspath(pathlike):
        return pathlike
```

This could be placed at the top of `loadsave.py` and `filebasedimages.py`, and inserted straightforwardly into the four functions/methods.

Reply:
I may be missing something, but python 3.4+ has the pathlib module built in, however `os.fspath` was only included in 3.6. So we will need a way to deal with pathlib.Path objects without `os.fspath` for py 3.4<3.6

Reply:
I see. I didn't fully appreciate the timeline here. Could we update the above suggestion to something like the following?

```Python
try:
    from os import fspath
except ImportError:
    def fspath(pathlike):
        try:
            pathlike = pathlike.__fspath__()
        except AttributeError:
            pass
        if not isinstance(pathlike, (bytes, unicode)):
            raise TypeError("expected str, bytes or os.PathLike object, not "
                            "{}".format(type(pathlike).__name__))
        return pathlike
```

This matches the docs:

```
os.fspath = fspath(path)
    Return the file system path representation of the object.
    
    If the object is str or bytes, then allow it to pass through as-is. If the
    object defines __fspath__(), then return the result of that method. All other
    types raise a TypeError.
```

It might go best in `nibabel.py3k`.

@fepegar @matthew-brett Feel free to chime in if you have cleaner/alternate suggestions.

Reply:
As I said in #612 this might be naive, but why not converting always to `str`?
```python
path = str(pathlike)
```
This feels like the simplest solution and it's perfectly backwards compatible. See https://github.com/python/cpython/blob/3.6/Lib/pathlib.py#L701-L702

Reply:
A few reasons to stay away from the `str(pathlike)` adaptation is justified in the `__fspath__` pep 519 https://www.python.org/dev/peps/pep-0519/#rationale. While calling `str` on pathlib.Path objects may work, calling `str` on `DirEntry` objects (as returned by `os.listdir` and `os.scandir`) does not work.

@effigies something I didn't see yesterday was that your suggestion still breaks on python 3.4<3.6 because unfortunately `os.fspath` and the dunder `.__fspath__()` were implemented in python 3.6+ so if we give it a python3.4 pathlib.Path object, it will hit the `TypeError` in your function.

In my opinion, `pandas` has done a pretty good job at resolving the pathlib issue. See https://github.com/pandas-dev/pandas/blob/53ad5c68b600f2929edd165c7b45e363c956144d/pandas/io/common.py#L121-L161

Reply:
That looks reasonable to me. I'm realizing py3k is a copy from numpy, so it doesn't make sense to add to that. Perhaps, following in the spirit of #530, we can start a utils.py for this function.

Pandas is BSD, so no problems including their code, but we will need to credit them.

Note also that we'll need to translate from their `pandas.compat.text_type` to `six.